File: open_mahjong_server/server/game_calculation/hand_structure.py
"""麻将手牌结构的通用定义。"""

from dataclasses import dataclass


@dataclass(frozen=True)
class HandStructure:
    """用固定面子数描述一类麻将的标准手牌结构。"""

    meld_count: int

    def __post_init__(self) -> None:
        if self.meld_count <= 0:
            raise ValueError("面子数必须为正数")

    @property
    def base_hand_tile_count(self) -> int:
        """未取得和牌张时的手牌张数。"""

        return self.meld_count * 3 + 1

    @property
    def complete_hand_tile_count(self) -> int:
        """取得和牌张后的手牌张数。"""

        return self.base_hand_tile_count + 1

    def concealed_meld_count(self, external_meld_count: int) -> int:
        """返回扣除副露后仍需由暗手组成的面子数。"""

        return self.meld_count - external_meld_count

    def concealed_tile_count(self, external_meld_count: int, *, complete: bool) -> int:
        """返回给定副露数下，暗手应有的标准张数。"""

        concealed_melds = self.concealed_meld_count(external_meld_count)
        return concealed_melds * 3 + (2 if complete else 1)


THIRTEEN_TILE_MAHJONG = HandStructure(meld_count=4)
SIXTEEN_TILE_MAHJONG = HandStructure(meld_count=5)

File: open_mahjong_unity/Assets/Scripts/GameScene/Calculation/CalculationScript/HandStructure.cs
using System;

/// <summary>用标准面子数描述一类麻将的手牌结构。</summary>
public sealed class HandStructure {
    public int MeldCount { get; }

    /// <summary>未取得和牌张时的标准手牌张数。</summary>
    public int BaseHandTileCount => MeldCount * 3 + 1;

    /// <summary>取得和牌张后的标准手牌张数。</summary>
    public int CompleteHandTileCount => BaseHandTileCount + 1;

    /// <summary>主手牌、一个牌宽的间隔和摸牌张合计占用的牌宽单位。</summary>
    public int DisplayWidthUnits => BaseHandTileCount + 2;

    public HandStructure(int meldCount) {
        if (meldCount <= 0) throw new ArgumentOutOfRangeException(nameof(meldCount));
        MeldCount = meldCount;
    }

    public int ConcealedMeldCount(int externalMeldCount) {
        return MeldCount - externalMeldCount;
    }

    public int ConcealedTileCount(int externalMeldCount, bool complete) {
        return ConcealedMeldCount(externalMeldCount) * 3 + (complete ? 2 : 1);
    }
}

/// <summary>客户端已支持的标准手牌结构及规则映射。</summary>
public static class HandStructures {
    public static readonly HandStructure ThirteenTile = new HandStructure(4);
    public static readonly HandStructure SixteenTile = new HandStructure(5);

    public static HandStructure Resolve(string roomRule, string subRule = null) {
        if (IsRule(roomRule, "taiwan") || IsRule(roomRule, "taiwan/standard")
            || IsRule(subRule, "taiwan/standard")) {
            return SixteenTile;
        }
        return ThirteenTile;
    }

    private static bool IsRule(string value, string expected) {
        return string.Equals(value, expected, StringComparison.OrdinalIgnoreCase);
    }
}
